BPA/Ch.10/js_tempran.stan

```diff
@@ -80,103 +149,52 @@ transformed data {
 
   // Constraints
   for (i in 1:M) {
     for (t in 1:(n_occasions - 1))
-      phi[i, t] <- inv_logit(logit(mean_phi) + epsilon[t]);
+      phi[i, t] = inv_logit(logit(mean_phi) + epsilon[t]);
     for (t in 1:n_occasions)
-      p[i, t] <- mean_p;
-  } //i
+      p[i, t] = mean_p;
+  }
 
   // Uncaptured probability
-  chi <- prob_uncaptured(M, n_occasions, p, phi);
+  chi = prob_uncaptured(M, n_occasions, p, phi);
 }
 
 model {
-  vector[n_occasions] qgamma;
-
-  qgamma <- 1.0 - gamma;
-
   // Priors
-  mean_phi ~ uniform(0, 1);
-  mean_p ~ uniform(0, 1);
-  gamma ~ uniform(0, 1);
+  // Uniform priors are implicitly defined.
+  //  mean_phi ~ uniform(0, 1);
+  //  mean_p ~ uniform(0, 1);
+  //  gamma ~ uniform(0, 1);
+  //  sigma ~ uniform(0, 5);
+  // In case of using a weakly informative prior
+  //  sigma ~ normal(2.5, 1.25);
   epsilon ~ normal(0, sigma);
-  sigma ~ uniform(0, 5);
 
   // Likelihood
-  for (i in 1:M) {
-    vector[n_occasions] qp;
-
-    qp <- 1.0 - p[i]';
-
-    if (first[i]) { // Captured
-      // Until first capture
-      if (first[i] == 1) {
-        1 ~ bernoulli(gamma[1] * p[i, 1]);
-      } else {
-        vector[first[i]] lp;
-
-        // Entered at 1st occasion
-        lp[1] <- bernoulli_log(1, gamma[1])
-          + bernoulli_log(1, prod(qp[1:(first[i] - 1)]))
-          + bernoulli_log(1, prod(phi[i, 1:(first[i] - 1)]))
-          + bernoulli_log(1, p[i, first[i]]);
-        // Entered at t-th occasion (1 < t < first[i])
-        for (t in 2:(first[i] - 1))
-          lp[t] <- bernoulli_log(1, prod(qgamma[1:(t - 1)]))
-            + bernoulli_log(1, gamma[t])
-            + bernoulli_log(1, prod(qp[t:(first[i] - 1)]))
-            + bernoulli_log(1, prod(phi[i, t:(first[i] - 1)]))
-            + bernoulli_log(1, p[i, first[i]]);
-        lp[first[i]] <- bernoulli_log(1, prod(qgamma[1:(first[i] - 1)]))
-          + bernoulli_log(1, gamma[first[i]])
-          + bernoulli_log(1, p[i, first[i]]);
-        increment_log_prob(log_sum_exp(lp));
-      }
-      // Until last capture
-      for (t in (first[i] + 1):last[i]) {
-        1 ~ bernoulli(phi[i, t - 1]);   // Survived
-        y[i, t] ~ bernoulli(p[i, t]);   // Capture/Non-capture
-      }
-      // Subsequent occasions
-      1 ~ bernoulli(chi[i, last[i]]);
-    } else {         // Never captured
-      vector[n_occasions+1] lp;
-
-      // Entered at 1st occasion, but never captured
-      lp[1] <- bernoulli_log(1, gamma[1])
-        + bernoulli_log(0, p[i, 1])
-        + bernoulli_log(1, chi[i, 1]);
-      // Entered at t-th occasion, but never captured
-      for (t in 2:n_occasions)
-        lp[t] <- bernoulli_log(1, prod(qgamma[1:(t - 1)]))
-          + bernoulli_log(1, gamma[t])
-          + bernoulli_log(0, p[i, t])
-          + bernoulli_log(1, chi[i, t]);
-      // Never entered
-      lp[n_occasions + 1] <- bernoulli_log(1, prod(qgamma));
-      increment_log_prob(log_sum_exp(lp));
-    }
-  }
+  jolly_seber_lp(M, n_occasions, y, first, last,
```

Comment:
If you named the function `jolly_seber_lpmf` you could write

```
M ~ jolly_seber(n_occassions, y, first, last, p, phi, gamma, chi);
```

No efficiency difference and which you find more natural is of course a matter of taste.  For the `jolly_seber_lpmf` version, you'd accumulate the densities as you go.  So it probably won't be quite as efficient as the way you have it because we don't yet have a way to drop constants in anything other than built-in (not user defined) sampling statements.